src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2713,16 +2714,56 @@ bool StorageReplicatedMergeTree::executeReplaceRange(LogEntry & entry)
                 .metadata_version_to_write = metadata_snapshot->getMetadataVersion()
             };
-            auto [res_part, temporary_part_lock] = cloneAndLoadDataPartOnSameDisk(
-                part_desc->src_table_part,
-                TMP_PREFIX + "clone_",
-                part_desc->new_part_info,
-                metadata_snapshot,
-                clone_params,
-                getContext()->getReadSettings(),
-                getContext()->getWriteSettings());
-            part_desc->res_part = std::move(res_part);
-            part_desc->temporary_part_lock = std::move(temporary_part_lock);
+
+            if (part_desc->src_table_part->info.partition_id != part_desc->new_part_info.partition_id)
+            {
+                auto src_part = part_desc->src_table_part;
+                const auto & src_storage = src_part->storage;
+
+                auto metadata_manager = std::make_shared<PartMetadataManagerOrdinary>(src_part.get());
+                IMergeTreeDataPart::MinMaxIndex min_max_index;
+
+                min_max_index.load(src_part->storage, metadata_manager);
+
+                MergeTreePartition new_partition;
+
+                new_partition.create(metadata_snapshot, min_max_index.getBlock(src_storage), 0u, getContext());
+
+                /// This will generate unique name in scope of current server process.
+                Int64 temp_index = insert_increment.get();
+
+                auto partition_id = new_partition.getID(*this);
+
+                MergeTreePartInfo dst_part_info(partition_id, temp_index, temp_index, src_part->info.level);
+
+                auto [res_part, temporary_part_lock] = cloneAndLoadPartOnSameDiskWithDifferentPartitionKey(
+                    src_part,
+                    TMP_PREFIX + "clone_",
+                    part_desc->new_part_info,
+                    metadata_snapshot,
+                    new_partition,
+                    min_max_index,
+                    clone_params,
+                    getContext()->getReadSettings(),
+                    getContext()->getWriteSettings());
+
+                part_desc->res_part = std::move(res_part);
+                part_desc->temporary_part_lock = std::move(temporary_part_lock);
+            }
+            else
+            {
+                auto [res_part, temporary_part_lock] = cloneAndLoadDataPartOnSameDisk(
+                    part_desc->src_table_part,
+                    TMP_PREFIX + "clone_",
+                    part_desc->new_part_info,
+                    metadata_snapshot,
+                    clone_params,
+                    getContext()->getReadSettings(),
+                    getContext()->getWriteSettings());
+
+                part_desc->res_part = std::move(res_part);
+                part_desc->temporary_part_lock = std::move(temporary_part_lock);
+            }
```

Comment:
It would be nice to avoid copy-paste (looks like it's exactly the same for MergeTree and ReplicatedMergeTree)

Reply:
Also, `insert_increment` doesn't make any sense for ReplicatedMergeTree because it uses sequential nodes in ZooKeeper

Reply:
@tavplubix 

Where do you suggest I put this code? Two new overloads for `MergeTreeData::cloneAndLoadDataPartOnSameDisk` and `MergeTreeData::cloneAndLoadPartOnSameDiskWithDifferentPartitionKey` or inside `MergeTreeDataPartCloner`. I think I would go with the former

Reply:
I don't have a strong opinion on this

Reply:
@tavplubix 

I have introduced a method in the base class that encapsulates the logic that preceeds `cloneAndLoadPartOnSameDiskWithDifferentPartitionKey`. This change removed some duplication, but not everything.

The decision of which method to call `cloneAndLoadDataPartOnSameDisk` or `cloneAndLoadPartOnSameDiskWithDifferentPartitionKey` still happens on caller site. The main reason for this was to avoid having a boolean parameter that would control internal behavior.

If you consider it a must, I can implement it.